File: app/src/main/java/com/arnaudpiroelle/manga/core/utils/FileHelper.java
package com.arnaudpiroelle.manga.core.utils;

import android.os.Environment;

import com.arnaudpiroelle.manga.model.Chapter;
import com.arnaudpiroelle.manga.model.Manga;
import com.arnaudpiroelle.manga.model.Page;

import java.io.File;

import javax.inject.Inject;

public class FileHelper {

    @Inject
    public FileHelper() {

    }

    public File getEbooksFolder() {
        return new File(Environment.getExternalStorageDirectory(), "eBooks");
    }

    public File getMangaFolder(Manga manga) {
        return new File(getEbooksFolder(), manga.getName());
    }

    public File getChapterFolder(Manga manga, Chapter chapter) {
        return new File(getMangaFolder(manga), chapter.getChapterNumber());
    }

    public File getPageFile(Manga manga, Chapter chapter, Page page) {
        String pageFormated = String.format("%03d.%s", Integer.valueOf(page.getPageNumber()), page.getExtension());

        File chapterFolder = getChapterFolder(manga, chapter);
        chapterFolder.mkdirs();

        return new File(chapterFolder + "/" + pageFormated);
    }
}

File: app/src/main/kotlin/com/arnaudpiroelle/manga/core/utils/FileHelper.kt
package com.arnaudpiroelle.manga.core.utils

import android.os.Environment.getExternalStorageDirectory
import com.arnaudpiroelle.manga.api.model.Chapter
import com.arnaudpiroelle.manga.api.model.Page
import com.arnaudpiroelle.manga.data.model.Manga
import java.io.File

open class FileHelper(val preferencesHelper: PreferencesHelper) {

    private fun getEbooksFolder(): File {
        val outputFolder = preferencesHelper.getOutputFolder()

        return if (outputFolder != null) {
            File(outputFolder)
        } else {
            File(getExternalStorageDirectory(), "eBooks")
        }
    }

    fun getMangaFolder(manga: Manga): File {
        return File(getEbooksFolder(), manga.name)
    }

    fun getChapterFolder(manga: Manga, chapter: Chapter): File {
        return File(getMangaFolder(manga), chapter.chapterNumber)
    }

    fun getPageFile(manga: Manga, chapter: Chapter, page: Page, pagePosition: Int): File {
        val pageFormated = "%03d.%s".format((pagePosition + 1), page.getExtension())

        val chapterFolder = getChapterFolder(manga, chapter)
        chapterFolder.mkdirs()

        return File(chapterFolder, pageFormated)
    }
}
